hbase-server/src/main/java/org/apache/hadoop/hbase/master/MetaRegionLocationCache.java

```diff
@@ -0,0 +1,217 @@
+ */
+package org.apache.hadoop.hbase.master;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentNavigableMap;
+import org.apache.hadoop.hbase.HRegionLocation;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.client.RegionInfoBuilder;
+import org.apache.hadoop.hbase.client.RegionReplicaUtil;
+import org.apache.hadoop.hbase.exceptions.DeserializationException;
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.types.CopyOnWriteArrayMap;
+import org.apache.hadoop.hbase.util.RetryCounter;
+import org.apache.hadoop.hbase.util.RetryCounterFactory;
+import org.apache.hadoop.hbase.zookeeper.ZKListener;
+import org.apache.hadoop.hbase.zookeeper.ZKUtil;
+import org.apache.hadoop.hbase.zookeeper.ZKWatcher;
+import org.apache.hadoop.hbase.zookeeper.ZNodePaths;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A cache of meta region location metadata. Registers a listener on ZK to track changes to the
+ * meta table znodes. Clients are expected to retry if the meta information is stale. This class
+ * is thread-safe.
+ */
+@InterfaceAudience.Private
+public class MetaRegionLocationCache extends ZKListener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MetaRegionLocationCache.class);
+
+  // Maximum number of times we retry when ZK operation times out.
+  private static final int MAX_ZK_META_FETCH_RETRIES = 10;
+  // Sleep interval ms between ZK operation retries.
+  private static final int SLEEP_INTERVAL_MS_BETWEEN_RETRIES = 1000;
+  private final RetryCounterFactory retryCounterFactory =
+      new RetryCounterFactory(MAX_ZK_META_FETCH_RETRIES, SLEEP_INTERVAL_MS_BETWEEN_RETRIES);
+
+  private final ZKWatcher watcher;
+  // Cached meta region locations indexed by replica ID.
+  // CopyOnWriteArrayMap ensures synchronization during updates and a consistent snapshot during
+  // client requests. Even though CopyOnWriteArrayMap copies the data structure for every write,
+  // that should be OK since the size of the list is often small and mutations are not too often
+  // and we do not need to block client requests while mutations are in progress.
+  private final CopyOnWriteArrayMap<Integer, HRegionLocation> cachedMetaLocations;
+
+  private enum ZNodeOpType {
+    INIT,
+    CREATED,
+    CHANGED,
+    DELETED
+  };
+
+  MetaRegionLocationCache(ZKWatcher zkWatcher) {
+    super(zkWatcher);
+    watcher = zkWatcher;
+    cachedMetaLocations = new CopyOnWriteArrayMap<>();
+    watcher.registerListener(this);
+    // Populate the initial snapshot of data from meta znodes.
+    // This is needed because stand-by masters can potentially start after the initial znode
+    // creation.
+    populateInitialMetaLocations();
+  }
+
+  private void populateInitialMetaLocations() {
```

Comment:
@ndimiduk / @xcangCRM Used RetryCounterFactory that also does exponential backoff. FYI.